File: leetcodes/15. 3Sum/3Sum.java
import java.util.Arrays;
import java.util.Collection;

class Solution {
  public List<List<Integer>> threeSum(int[] nums) {
    //input: array
    //output: array with triplet
    //edge case: if len < 3, return empty list
    //use SET to avoid duplicates
    //sort the original array
    //iterate over the array
      //start from one element
      //if nums[i] > 0, break;
      //use while loop to adjust two pointers
        //check if sum is 0 then add to the list
          //same i, j++, k--
        //else if sum <0 , j++ (until big enough )
        //else if sum > 0, k-- (until small enough)
    //return new list
    //time O(n^2)(for + while > sort O(logn)), space O(n) (hashset?)
    Set<List<Integer>> res = new HashSet<>();
    if (nums.length < 3) return new ArrayList<>(res);
    Arrays.sort(nums);
    for(int i = 0; i < nums.length - 2; i++) {
      int j = i + 1;
      int k = nums.length - 1;
      if (nums[i] > 0) break;
      while (j < k) {
        int sum = nums[i] + nums[j] + nums[k];
        if (sum == 0) res.add(Arrays.asList(nums[i], nums[j++], nums[k--]));
        else if (sum > 0) k--;
        else if (sum < 0) j++;
      }
    }
    return new ArrayList<>(res);
  }
}

File: leetcodes/15. 3Sum/3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 var threeSum = function(nums) {
  //two pointers: sort nums
   //i, l, r -> itearte over the nums array (i from 0 to len - 3 to have at least 3 elem) 
  // while(l < r) l = i + 1, r = len - 1, i + l + r = s
  // if s < 0 -> l++
  // else if s > 0  -> r--
  // if s == 0-> push to result array ,  l++, r--
  //return res array
  //time O(n^3), space O(n)
const res = [] 
if (nums.length < 3) return res;
nums.sort((a, b) => a - b)

for (let i =0; i < nums.length - 2; i++) {
  let l = i +1;
  let r = nums.length - 1;
  if (nums[i]>0) break;
  if (i >0 && nums[i] === nums[i-1]) continue;  
  while(l < r) {
    let s = nums[i] + nums[l] + nums[r]
    if (s > 0) r--;
    else if (s < 0) l++;
    //has to reorder to avoid OLE (if I can use set, should be better)
    else {
      res.push([nums[i], nums[l], nums[r]])
       while((l<r) && (nums[l] === nums[l + 1])) l++;
       while((l<r) && (nums[r] === nums[r - 1])) r--; 
       l++;
       r--;
    }
  }
}
return res;
};